src/lib/Canvas.js

```diff
@@ -0,0 +1,762 @@
+        if (this.buffer.width > 0 && this.buffer.height > 0) {
+            this.canvasCTX.drawImage(this.buffer, 0, 0);
+        }
+    },
+
+    dispatchNewEvent: function(event, name, detail) {
+        detail = {
+            detail: detail || {}
+        };
+        detail.detail.primitiveEvent = event;
+        return this.canvas.dispatchEvent(new CustomEvent(name, detail));
+    },
+
+    dispatchNewMouseKeysEvent: function(event, name, detail) {
+        detail = detail || {};
+        detail.mouse = this.mouseLocation;
+        detail.keys = this.currentKeys;
+        return this.dispatchNewEvent(event, name, detail);
+    },
+
+    finmousemove: function(e) {
+        if (!this.isDragging() && this.mousedown) {
+            this.beDragging();
+            this.dispatchNewMouseKeysEvent(e, 'fin-canvas-dragstart', {
+                isRightClick: this.isRightClick(e)
+            });
+            this.dragstart = new rectangular.Point(this.mouseLocation.x, this.mouseLocation.y);
+        }
+        this.mouseLocation = this.getLocal(e);
+        //console.log(this.mouseLocation);
+        if (this.isDragging()) {
+            this.dispatchNewMouseKeysEvent(e, 'fin-canvas-drag', {
+                dragstart: this.dragstart,
+                isRightClick: this.isRightClick(e)
+            });
+        }
+        if (this.bounds.contains(this.mouseLocation)) {
+            this.dispatchNewMouseKeysEvent(e, 'fin-canvas-mousemove');
+        }
+    },
+
+    finmousedown: function(e) {
+        this.mouseLocation = this.mouseDownLocation = this.getLocal(e);
+        this.mousedown = true;
+
+        this.dispatchNewMouseKeysEvent(e, 'fin-canvas-mousedown', {
+            isRightClick: this.isRightClick(e)
+        });
+        this.takeFocus();
+    },
+
+    finmouseup: function(e) {
+        if (this.isDragging()) {
+            this.dispatchNewMouseKeysEvent(e, 'fin-canvas-dragend', {
+                dragstart: this.dragstart,
+                isRightClick: this.isRightClick(e)
+            });
+            this.beNotDragging();
+            this.dragEndtime = Date.now();
+        }
+        this.mousedown = false;
+        this.dispatchNewMouseKeysEvent(e, 'fin-canvas-mouseup', {
+            isRightClick: this.isRightClick(e)
+        });
+        //this.mouseLocation = new rectangular.Point(-1, -1);
+    },
+
+    finmouseout: function(e) {
+        if (!this.mousedown) {
+            this.mouseLocation = new rectangular.Point(-1, -1);
+        }
+        this.dispatchNewMouseKeysEvent(e, 'fin-canvas-mouseout');
+    },
+
+    finwheelmoved: function(e) {
+        if (this.isDragging() || !this.hasFocus()) {
+            return;
+        }
+        e.preventDefault();
+        this.dispatchNewMouseKeysEvent(e, 'fin-canvas-wheelmoved', {
+            isRightClick: this.isRightClick(e)
+        });
+    },
+
+    finclick: function(e) {
+        var delay = this.component.resolveProperty('doubleClickDelay');
+        if (delay < 100) {
```

Comment:
Yes, 100ms is the low end of the range for double clicking. Anything less than that is ridiculous.